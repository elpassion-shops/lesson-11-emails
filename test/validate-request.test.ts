import {
  IsEmail,
  IsInt,
  IsString,
  Max,
  Min,
  validate,
  ValidateNested,
} from "class-validator";

class Answer {
  @IsString()
  id!: string;

  @IsInt()
  @Min(1)
  @Max(5)
  choice!: number;
}

class Request {
  @IsString()
  @IsEmail()
  email!: string;

  @ValidateNested({ each: true })
  answers!: Answer[];
}

function generateRequest(overrides: { email?: any; answers?: any }) {
  const request = new Request();
  request.email =
    overrides.email === undefined ? "mihau@gmail.com" : overrides.email;
  request.answers = overrides.answers === undefined ? [] : overrides.answers;

  return request;
}

test("passes with correct request", async () => {
  const request = generateRequest({});
  expect(await validate(request)).toEqual([]);
});

test("fails without email", async () => {
  const request = generateRequest({ email: null });
  expect(await validate(request)).toEqual([
    expect.objectContaining({
      constraints: {
        isEmail: "email must be an email",
        isString: "email must be a string",
      },
      property: "email",
    }),
  ]);
});

test("fails with incorrect email", async () => {
  const request = generateRequest({ email: "mihaugmail.com" });
  expect(await validate(request)).toEqual([
    expect.objectContaining({
      constraints: {
        isEmail: "email must be an email",
      },
      property: "email",
    }),
  ]);
});

test("fails with empty answers", async () => {
  const request = generateRequest({ answers: null });
  expect(await validate(request)).toEqual([
    expect.objectContaining({
      children: [
        expect.objectContaining({
          constraints: {
            nestedValidation:
              "each value in nested property answers must be either object or array",
          },
          property: "answers",
        }),
      ],
    }),
  ]);
});

test("fails with incorrect answer", async () => {
  const request = generateRequest({ answers: [new Answer()] });
  expect(await validate(request)).toMatchSnapshot();
});
