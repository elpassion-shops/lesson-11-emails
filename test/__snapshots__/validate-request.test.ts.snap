// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fails with answer without id 1`] = `
Array [
  ValidationError {
    "children": Array [
      ValidationError {
        "children": Array [
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "isString": "id must be a string",
            },
            "property": "id",
            "target": Answer {},
            "value": undefined,
          },
        ],
        "property": "0",
        "target": Array [
          Answer {},
        ],
        "value": Answer {},
      },
    ],
    "property": "answers",
    "target": Request {
      "answers": Array [
        Answer {},
      ],
      "email": "mihau@gmail.com",
    },
    "value": Array [
      Answer {},
    ],
  },
]
`;

exports[`fails with incorect answer 1`] = `
Array [
  ValidationError {
    "children": Array [
      ValidationError {
        "children": Array [
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "isString": "id must be a string",
            },
            "property": "id",
            "target": Answer {},
            "value": undefined,
          },
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "isInt": "choice must be an integer number",
              "max": "choice must not be greater than 5",
              "min": "choice must not be less than 1",
            },
            "property": "choice",
            "target": Answer {},
            "value": undefined,
          },
        ],
        "property": "0",
        "target": Array [
          Answer {},
        ],
        "value": Answer {},
      },
    ],
    "property": "answers",
    "target": Request {
      "answers": Array [
        Answer {},
      ],
      "email": "mihau@gmail.com",
    },
    "value": Array [
      Answer {},
    ],
  },
]
`;

exports[`fails with incorrect answer 1`] = `
Array [
  ValidationError {
    "children": Array [
      ValidationError {
        "children": Array [
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "isString": "id must be a string",
            },
            "property": "id",
            "target": Answer {},
            "value": undefined,
          },
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "isInt": "choice must be an integer number",
              "max": "choice must not be greater than 5",
              "min": "choice must not be less than 1",
            },
            "property": "choice",
            "target": Answer {},
            "value": undefined,
          },
        ],
        "property": "0",
        "target": Array [
          Answer {},
        ],
        "value": Answer {},
      },
    ],
    "property": "answers",
    "target": Request {
      "answers": Array [
        Answer {},
      ],
      "email": "mihau@gmail.com",
    },
    "value": Array [
      Answer {},
    ],
  },
]
`;
